### ISSUE TITLE
MonoGame on Web

### ISSUE THREAD

**Issue by cra0zy** [mono/MonoGame#3895]
I've been thinking on doing some work on this, even tho there is base for JSIL I would like to think of other options as well. Current options on getting C# on Web are:
 - JSIL  (may translate C# badly in some cases, but it does produce files runnable even from android)
 - NaCl  (requires browser plugin, but will be more precise in running C# than JSIL)
 - PNaCl  (Chrome only, no plugin required)

@tomspilman @KonajuGames @dellis1972 What are your thoughts on this?

**Comment by WardBenjamin** [comment]
I've done a little research into JSIL, and found something that I think you guys will find very interesting.

http://hildr.luminance.org/MonoGame/Platformer2D/

It looks like somebody (looks like a contributer to JSIL, actually) decided to do this already. His code can be found at https://github.com/kg/MonoGame/tree/JSIL. The fact that this produces something that we can run on android is also really fantastic.

**Comment by cra0zy** [comment]
@kg is the creator of JSIL....

Also in the main JSIL repo there is some code dedicated to porting MonoGame to web(https://github.com/sq/JSIL/tree/master/Libraries/MonoGame), but the way MonoGame planned to do it is a bit different(create a solution for MonoGame).

Also @WardBenjamin do Protobuild.exe --generate Web I think, it will generate JSIL web stuff MonoGame already has.

**Comment by WardBenjamin** [comment]
> @kg is the creator of JSIL....
I didn't know that :)

I'm going to check that out, thanks!

**Comment by kg** [comment]
A client sponsored some recent work on MonoGame & SDL2 (via p/invoke and emscripten) integration for JSIL that is almost ready to merge. I'll try to remember to mention that here when it's available.

On May 30, 2015 8:59:01 AM PDT, Harry <notifications@github.com> wrote:
>@kg is the creator of JSIL....
>
>Also in the main JSIL repo there is some code dedicated to porting
>MonoGame to
>web(https://github.com/sq/JSIL/tree/master/Libraries/MonoGame), but the
>way MonoGame planned to do it is a bit different(create a solution for
>MonoGame).
>
>Also @WardBenjamin do Protobuild.exe --generate Web I think, it will
>generate JSIL web stuff MonoGame already has.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mono/MonoGame/issues/3895#issuecomment-107056912

- kg (mobile)

**Comment by cra0zy** [comment]
@kg Sweet!!!

**Comment by tomspilman** [comment]
> NaCl
> PNaCl 

I don't think either of these are good solutions as they are specific to just one or two browsers.  The only path I see that works on all browsers and will be fairly fast on most of them is cross-compiling to JavaScript.

This is why we decided on JSIL in the first place.

**Comment by cra0zy** [comment]
Ok, I'm going to wait up to a month to see what's going to happen with what @kg talked about. After that I'll start working on this using JSIL.

**Comment by cra0zy** [comment]
http://www.godotengine.org/wp/webassembly-godot-on-the-web/

**Comment by tomspilman** [comment]
It will be interesting to see when WebAssembly becomes usable...  maybe a year or two if we're lucky.  Until then JSIL is still a good path for us.

**Comment by kg** [comment]
At some point you'll be able to use WebAssembly with compilers like JSIL and GWT, hopefully. It's on the long-term agenda.

**Comment by kg** [comment]
The initial monogame work is almost complete (been doing it in my spare time and paying out of pocket since the sponsor bailed), so I'll have a demo & some associated documentation on the JSIL wiki out soon for people to use as a starting point. It's the MonoGame version of the platformer2d sample, running against a fork and using emscripten'd SDL2. Gamepad input, audio, opengl, etc all work. Getting something like this working in trunk monogame will be a challenge but once this is polished up it will be a good proof-of-concept for how to do it.

**Comment by kg** [comment]
Here's the working demo.
http://hildr.luminance.org/SampleFNA/
https://github.com/sq/SampleFNA

It's using FNA instead of trunk MonoGame, because FNA's use of SDL2 as its single backend made the initial port easier. The general approach should be viable for MG, since it lets you keep all of your PInvoke-based code (that talks to GL or whatever). It's an approach that's also compatible with writing a separate web backend that reuses a subset of the existing native code, instead of reusing all of it and patching up specific spots with JS.

There are a handful of issues left to fix but it's a pretty good proof-of-concept for something like this.

**Comment by SamuelEnglard** [comment]
Crashes in IE11 on Win 8.1

    Unhandled exception at http://hildr.luminance.org/SampleFNA/bin/JSIL/FNA,%20Version=0.0.0.1,%20Culture=neutral,%20PublicKeyToken=null.js line 11486:
    Unable to get property 'bufferSubData' of undefined or null reference

**Comment by kg** [comment]
Last time I tested it, IE's OpenGL support was too old, check the error console. If it's something other than GL support I can fix it.

FWIW, that is a major downside to the approach of using emscripten-compiled SDL2 or otherwise: Existing graphics stacks are still pretty heavily dependent on a full GL feature set, and WebGL is more of a stripped-down, crippled GLES.

**Comment by SamuelEnglard** [comment]
Guessing WebGL 0.94 is too old. Aww :(

**Comment by cra0zy** [comment]
@kg Just FYI about:

> For Song playback to work, you'll need to convert Music.wma into an ogg vorbis file called Music.ogg, next to the Music.xnb file created by the content builder above. Sadly, the content builder cannot do this for us. (JSILc may acquire the ability to do this in the future.)

Is not correct, the newest develop branch Pipeline Tool will convert song files to .ogg files as long as platform is set to DesktopGL / WindowsGL / Linux and processor is set to Song.

**Comment by cra0zy** [comment]
While setting up content manager with JSIL I found that: https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Content/ContentManager.cs#L200 causes a crash.

It's not that big of a deal since I was not gonna use that part of the code anyway, but I thought it would be useful to mention it.

@kg For Xna/Fna are you loading xnb or raw(png, ogg, etc)? I started setting up raw loading of stuff.

**Comment by kg** [comment]
Fna loads raw xnbs and we kept that intact, with the exception of streaming music - I used the browser's audio decoder for that. You could call out to libvorbis I suppose but you'd need to do the emscripten port of that thing yourself.

My hand-written xna stubs used the browser's image decoders (jpeg/png etc) but that's a really bad idea as browsers tend to have really buggy/inconsistent ones.

FYI, please file JSIL bug reports about crashes that only occur in JS builds.

**Comment by cra0zy** [comment]
Wow you are fast.

> FYI, please file JSIL bug reports about crashes that only occur in JS builds.

I will, first I need to figure out a small code that can reproduce it.

**Comment by ashes999** [comment]
Is this still supported?

@kg your demo link crashes under the latest Chrome:

```
Unhandled exception at http://hildr.luminance.org/SampleFNA/SDL2.js line 4:
Uncaught TypeError: Cannot read property 'getParameter' of undefined
```

**Comment by kg** [comment]
@ashes999 Works here, just tried it. Browser vendors love to break HTML5 games, so maybe file a bug report against Chrome. SDL2.js is an emscripten-compiled version of SDL2.

**Comment by ashes999** [comment]
@kg I'm running Chrome `50.0.2661.102`. How about you?

This doesn't exactly inspire me with confidence. I can imagine a fragmented user base complaining about my game crashing and/or not working, and being stuck without the ability to easily troubleshoot it.

Is there a well-defined subset that this should work with? Can I say something like "always works with the latest browser(s)"?

**Comment by kg** [comment]
It probably requires WebGL and you don't have it. SDL2.js isn't something I wrote, so it's not under my control.

> This doesn't exactly inspire me with confidence. I can imagine a fragmented user base complaining about my game crashing and/or not working, and being stuck without the ability to easily troubleshoot it.

Welcome to HTML5. It's not great.

**Comment by cra0zy** [comment]
Just gonna mention that MonoGame has shifted to using SDL instead of OpenTK so emscripten-compiled version may be an option for MonoGame as well.

**Comment by ashes999** [comment]
@kg I worked directly on JS games in the past and experienced something similar. Thanks for clarifying. I think this is a pretty viable approach for now.

**Comment by cra0zy** [comment]
Only input is implemented, graphics/audio/loading content is not.

**Comment by SamuelEnglard** [comment]
In theory they could give my MonoGameJs a try?

**Comment by RiuSalvi** [comment]
Any news about c#, monogame and web?

I would like to run my monogame projects in a browser and i do not know how.

**Comment by bencrossman** [comment]
I would be interested in any updates too

**Comment by cra0zy** [comment]
@bencrossman You have some psychic powers my friend :), yesterday I started working on MG Bridge.net implementation, most of the things aren't yet implemented, but I did get it to compile and run at least:

![bridgenet](https://user-images.githubusercontent.com/6773302/30836199-e5b08c36-a25c-11e7-9d38-5c0c34e4f268.png)

PS. I know about web assembly, I just have my personal reasons for wanting to do it this way.

**Comment by bencrossman** [comment]
That's awesome. Well done

**Comment by tomspilman** [comment]
> PS. I know about web assembly

WebAssembly for C# to JS still seems to be in development and not ready for our use, but i expect we will eventually support it.

**Comment by cra0zy** [comment]
Status update: Mouse, Keyboard and Song implemented.

**Comment by TheSpydog** [comment]
@cra0zy Any update on your Bridge.net implementation? It looks really exciting! :smile:

**Comment by cra0zy** [comment]
Nah, waiting for webassembly to mature. I could get Bridge.Net version working, but it requires a lot of work and I don't have any incentive to do so.

**Comment by cra0zy** [comment]
> @cra0zy - Did you look at Blazor?

Its meant more for writing html applications so to speak, it does not offer any good way of invoking a lot of stuff we need (like webgl for example).

**Comment by tomspilman** [comment]
>  it does not offer any good way of invoking a lot of stuff we need (like webgl for example).

What would be a "good way"?  They say it has support for invoking Javascript from C#.

**Comment by cretz** [comment]
A good way is something like https://github.com/rustwasm/wasm-bindgen for .Net. Also note that Mono's WASM support is just the Mono interpreter emscripten'd. Your C# is not AOT compiled to WASM as might be expected (though that's a goal for them).

**Comment by cra0zy** [comment]
> What would be a "good way"? They say it has support for invoking Javascript from C#.

Doing that per each call gives slow performance.

**Comment by samuelGrahame** [comment]
Yeah I was playing around with Blazor. In my opinion until web assembly can access things like the dom/js api without javascript. best solution is using bridge - c# to JavaScript.

But I didn't see a issue posted in Blazor : https://github.com/aspnet/Blazor/issues/1212

they were talking about maybe inclining js using bridge/ retyped.

**Comment by cra0zy** [comment]
https://twitter.com/cra0zy/status/1031188387783475200

**Comment by cra0zy** [comment]
http://community.monogame.net/t/monogame-inside-your-web-browser/10918

**Comment by LukeTOBrien** [comment]
Hello @cra0zy,

I've been interested in your thread and the idea of running MonoGame inside the browser, what do you think of Web Assembly and [Blazor](https://blazor.net/)?  
I think Web Assembly would be really ideal for MonoGame, DLLs can be loaded rather than having to generate JS.

I am also interested in [Web VR](https://webvr.info/) and I think MonoGame could be used for that in the future.

What do ya think?

**Comment by cra0zy** [comment]
> I've been interested in your thread and the idea of running MonoGame inside the browser, what do you think of Web Assembly and [Blazor](https://blazor.net/)?

WebAssembly is way too experimental for C#, Blazor is slower than Bridge.NET.

> I think Web Assembly would be really ideal for MonoGame, DLLs can be loaded rather than having to generate JS.

That would be a horrible way to compile a C# game for web, TLDR. is you want to convert C# to WebAssembly, not make the C# runtime available on web, ie. Unity for example uses IL2CPP + Emscripten.

**Comment by samuelGrahame** [comment]
WebAssembly needs a GC -> Access to DOM -> Access to -> WebGL

**Comment by codec-abc** [comment]
@cra0zy : It seems you have benchmarked Blazor vs Bridge.Net regarding speed execution. Would you mind sharing it please?
Also, why packaging MonoGame as compiled assemblies, instead of truly "native" wasm modules, would be horrible? The later would probably be faster to load (no need for the Mono VM to be sent) and faster to execute (truly wasm, so less indirection). But besides this, are there more drawbacks to this approach?

**Comment by cra0zy** [comment]
To answer both your questions, I have not benchmarked it as it would be redundant to do so considering you would have to make a mandatory copy of each vertex/index array you pass to webgl + you would have to call javascript code multiple times just to get access to WebGL/Input/Audio/etc.

**Comment by Jjagg** [comment]
There's more nuance tot this than vertex copies I think. Would be neat if anyone got MG running on Blazor for comparing perf.

**Comment by LukeTOBrien** [comment]
At the moment I believe wasm was released as a MVP at version one, so it
doesn't have a lot of the features needed.
I have also heard that wasm is slower, at the moment, than JavaScript, and
the Blazor startup project is slower than I would expect from a pure
JavaScript one,
As wasm is still version 1 and doesn't have the supported features, a
benchmark would be misleading at this time.

But I do think it's something to keep an eye on, I think wasm will develop
more the the feature needed and probably will become a better option than
JavaScript in the future.

**Comment by cra0zy** [comment]
> There's more nuance to this than vertex copies I think.

You need to invoke JS to access WebGL/Audio/Input :/

**Comment by codec-abc** [comment]
How slow is Blazor compared to JS? if this something like 2x or even 3x times slower it can still be useful and would probably improve over time. Also, at least in Firefox wasm <-> Js calls [are now quite fast](https://hacks.mozilla.org/2018/10/calls-between-javascript-and-webassembly-are-finally-fast-%F0%9F%8E%89/).
So a first version of MG running with blazor isn't gonna beat JS right now in term of speed and app size, but it doesn't mean it would be bad. Actually, I would be surprised if performance is so terrible that it would be unusable.

**Comment by LukeTOBrien** [comment]
I think also we need MonoGame that is targeted toward .Net Standard -
Perhaps the 3.8 release?.

Blazor is a ASP.NET project and uses .Net Core, at the moment it uses Mono
to compile to wasm but in the future that could switch to CoreRT.

**Comment by codec-abc** [comment]
Cool!
Interesting result. I ran the Blazor project and had the same result. Interestingly enough the major performance hit seems to happen only in wasm code. Not even when going from Js to wasm nor GC. 
![image](https://user-images.githubusercontent.com/1208646/51147038-1c738700-1859-11e9-920e-8e77368fe120.png) So I am wondering what kind of code could result in so long operations.

**Comment by samuelGrahame** [comment]
I am going to covert to the latest version of blazor - currently running 0.4. they have 0.7

**Comment by codec-abc** [comment]
I have done it for a even simpler page where I have a div that take the full page and update a line of text on each `onmousemove` event. I expected the processing to be near 0 for the wasm operation and only have repaint operations (and maybe layout). It turns out that my CPU is not idle most of the time but only ~30% of the time. Repaint take around ~6%. Wasm account for ~35% of the time which is huge! I didn't expected Blazor to perform so poorly performance wise. I wonder what is responsible for such low performance.

**Comment by codec-abc** [comment]
This seems confirmed by this [Reddit thread](https://www.reddit.com/r/csharp/comments/8gjx3v/a_small_performance_comparison_of_monowasmblazor/) As such MonoGame would probably run terribly except for some toy applications. On the other hand, the day C# can be compiled directly to wasm (without needing the mono VM) performance could be nice enough to be used.

**Comment by kg** [comment]
Mono wasm is compiling C# to wasm, you're not losing time there unless you fall back to the interpreter (this should only happen in rare cases). That reddit thread is from 8 months ago so results that old should be treated as meaningless for a runtime in early development like this.

If you're seeing bad performance the only way to truly understand it is profiling, but in general some info:
* Cross-language inlining (wasm <-> js) is a thing in Firefox now and might also be available in V8 soon (I'm not certain there). I expect JSC will have it eventually too. So calling between environments is not implicitly a problem, as long as the JS bridging functions are cheap.
* The biggest interop hurdle you need to worry about is things being copied into and out of the wasm heap. It's unavoidable to avoid copies to/from the heap for JS strings, because JS strings have a bad API. For things like vertex and index buffers this should not be a problem - you can create ArrayViews that point into the wasm heap and then pass those to WebGL APIs when performing uploads. If you're manually copying in/out you're simply not bridging to WebGL correctly. If you want to marshal structs in/out of wasm you will eat a cost there too (either copying into JS objects or indirecting with properties).
* Some of these interop hurdles apply to Bridge (and JSIL, back before I stopped maintaining it) just like Wasm. If your vertex buffer is filled with structs you need some marshaling logic to actually turn it into something you can upload into WebGL. I don't know how Bridge solves that but JSIL solved it by using marshaling wrappers that treated the typed array as permanent storage for struct properties, which is complicated - and copying struct objects into a buffer manually once per vertex is also a slow, awful approach. Wasm-compiled C# will not have this problem, memory is memory.
* Passing JS objects around (and across the JS<->wasm boundary, in particular) is expensive and will be until host objects (anyref) are shipping in every wasm runtime. This is because you have to maintain an id<->object table on the JS side so that you can pass numeric object handles into/out of wasm. In many cases though you don't need that many host objects - MonoGame isn't going to be using a ton of them. Test cases like 'render a bunch of dynamic HTML' are the opposite of a good benchmark here because they rely heavily on both host objects and calls into non-JS APIs.
* In both cases (bridge and wasm) you need to actually profile your test case and see why it's slow. It's often a result of making a silly mistake or of a bug in one of the libraries you're using, and not a problem with the compiler or runtime. Using something like react or jquery from wasm is likely to perform very badly because react isn't designed for a runtime environment where strings and objects have high interop cost, but it would be possible to design a virtual DOM API for wasm that would be high-performance by making use of the wasm heap directly.

Anyway, in conclusion I think it is possible for both Bridge and WASM versions of MonoGame to perform okay right now, and in the long term WASM will wipe the floor with Bridge because you are never going to get native-level computational performance out of JS for things like vertex buffers and stuff. JS runtimes simply aren't smart or consistent enough to optimize that code for you (which is one of the reasons JSIL ended up being a dead end).

Context: I was one of the original designers of WebAssembly and I review some of the pull requests for mono's wasm support.

**Comment by codec-abc** [comment]
Thanks for the detailed answer. In case you want to take a look [I created a repository for a analyzing performances.](https://github.com/codec-abc/BlazorPerformanceTest) for the case I talked earlier: moving the mouse over an area and showing the move position by updating text inside a div. On both Firefox and Chrome the CPU usage is very high for a simple application like this. In the last commit I even removed the DOM update (so basically I just subscribe to an event and do nothing when it is fired) and the CPU usage is still quite high. I agree with you on profiling but analyzing time spent inside wasm code isn't easy. This is what I got:
![image](https://user-images.githubusercontent.com/1208646/51150925-e2aa7c80-1868-11e9-921d-b871c1f042e6.png) and this isn't giving much hint about what is wrong (assuming there is something wrong).

**Comment by Jjagg** [comment]
Thanks so much for that explanation @kg.

**Comment by cra0zy** [comment]
> * , but it would be possible to design a virtual DOM API for wasm that would be high-performance by making use of the wasm heap directly.

@kg If you don't mind, can you give me a simple example of calling a DOM related function in the most effective manner?

I know about: https://learn-blazor.com/architecture/interop/

**Comment by codec-abc** [comment]
@kg: I found a more up to date [benchmark](https://github.com/stefan-schweiger/bakkBenchmark) (September 2018) and an issue from the [blazor repo](https://github.com/aspnet/Blazor/issues/767) where member of the Blazor team say that saying a 100x time slower code (comparatively to desktop) is pretty standard at this point. Also, it doesn't seems the team focused on performance lately, because it is working well enough for their uses cases. Also, the ongoing work to get full AoD wasm compilation with Mono will certainly improve the situation. So my point is is that MonoGame would probably run terribly today and probably won't be usable for much things. On the other hand, getting something that work with mono-wasm today would allow developer to take advantage has it as soon as the mono-wasm AoT wasm compilation will land.

**Comment by harry-cpp** [comment]
![webassembly](https://user-images.githubusercontent.com/6773302/76265353-97b9a880-6264-11ea-99e7-37405510cc9b.png)

**Comment by codec-abc** [comment]
@harry-cpp : Nice! Is MonoGame close to be WASM compatible?

**Comment by harry-cpp** [comment]
Yes, yes it is.

**Comment by skyne98** [comment]
Would be incredible to see a Monogame instance as a component for Blazor.

**Comment by iamthefaker** [comment]
Could anyone write a paragraph dummies guide for how to enable an existing monogame project so that it can be deployed to the web? (Currently we are deploying to Windows Android Linux Mac and iOS ... it would be really helpful to be able to deploy to web browser. (Oh and probably Occulus Quest2 as well, but that's another story.)

**Comment by japhib** [comment]
Any update on the progress here?

Also @harry-cpp is the code for the web branch of MonoGame accessible anywhere? Would love to poke around and see what changes you've had to make

**Comment by lucasteles** [comment]
Would be great to have a monogame web template for the dotnet cli

**Comment by mattkanwisher** [comment]
.NET 6 has a ton of updates for WASM support, including native dependencies. Was wondering if anyone has looked at using monogame now since it seems like underlying infrastructure is a lot better?

**Comment by mrhelmut** [comment]
The web target is on hold, we lack the bandwidth to look into it. Contributions are more than welcome.

**Comment by codec-abc** [comment]
Could something like obtaining a grant from Microsoft could help setting paid bounty to get MonoGame running on WASM? Or maybe get someone from MS to help. Having MonoGame would be such a good showcase for the .Net WASM story they are developing it would make sense for them to spare some resources to get this running.

**Comment by ohtrobinson** [comment]
I believe nkast is working on getting MG working under WASM. The progress made so far is promising

**Comment by damian-666** [comment]
well Silverlight was part of that .net story and look what happened there.
 Flash, dead too.  Windows phone was a tragedy.   Netcore 6 tho i think is
a triumph.  Windows on ARM has hope too... But I think any high performance
web technology will get blocked eventually for security reasons or business
model, browser wars,  or whatever, so frankly i wouldnt risk going that way
as  i got burned before using silverlight too long.. i'll stick with the
app stores and the like even if the graphics took mae a year to
reimplement.  im glad i didn't spend too long on UWP either it could be
dropped and because  u can wrap a normal windows EXE now, easily as an app
package, and besides, the MS app store isn't really doing well either.
Whats great about steam, and net6 desktop,  is that u can someone can buy
ur game on a Windows,  and then buy a Mac or linux and still own it.

**Comment by iSeiryu** [comment]
Curious, if it's still promising. The last articles/posts I've seen about it were almost 2 years old.

**Comment by damian-666** [comment]
"This comment was marked as disruptive content.   "     you do this again i will no longer support or sponsor monogame.      Sensor me you have no idea who you are talknig to  or how deep the pockets are.     TOxic culture here don t expecgt any help from me or my billionare connections.

**Comment by damian-666** [comment]
@ISeiryu this h.as nothng to do with your quesetion,  i just got a notice and scroll up see my future predicting helpful comment indented to save people YEARS of wasted time, marked as "dispruptive"...     WTF marked it that? who has the right to censor someone else for trying to share his "life story"... to help for free when i get 200K + saslaray in Silicon valeey wiht20+  industry experience.   jsut a general comment about the culture and its #offtic and i dont care to put it where it belongs.

Even rocket girl shoul NOT  have been silenced  for asking the type of difficult questoins that people with deep pockets ask before they invest in a technology,  only to be met with dismissive and curt responses as was I went i asked about netstandard2. i it cost my comany 6 weeks, i deserved more thatn "not supported closed"  then she went off into a rage.     you have no idea who she is that would be funny if you did.    how many femailes actually show up here or contirbute?    zero .. her offer was the best most of you are gonig to get from a  real female... LOL.. i make a joke about it and get banned for a week for swearing lol  its this an elementary school...   every company i worked  in Silicon value,  the halls ans campuse are all profane outbursts of frustration along with the occasional  and rrare dance of success.    divesityty?   bunch of grumpy old dudes , toxic culture.    disresptful  of people who try to hlep and code of conduct does not exist , be nice , be respectufl by default. and tolerate.  would  be a start   Rocketgril puts on some heat and she gets disset.. now wonder she lauches and rage quits.      mark thsi as disruptive , ban me i dont care , i'm done "sharing" and being disrespected for actually promoting praising,  this group and its work and its focus.   i got what i need for now and focus on CPU and my own problems and those of industry leaders i work with .   bye.

**Comment by Beyley** [comment]
@damian-666 
Gonna try to keep this civil no matter how much i want to be mean/insulting here, given you are insulting me and probably many others with this comment.
Your comment was marked as disruptive as it is negative in nature, which is generally a no-go on github (if you want to complain about the state of the web, you should join the MonoGame Discord and complain in #off-topic, we often talk about that kind of stuff there). The point of this github issue is to fester a positive discussion about the problem, and work towards a solution (which nkast has been doing, they have some really cool progress!)
WASM games on the web are not dead, as WASM runtimes in browsers are improving, and becoming more secure, along with the tooling available for WASM increasing in usability. WebGPU is also showing itself to be a much needed successor to WebGL, and once released in stable on major browsers, may start a resurgance of web-based games.
The MonoGame community is not perfect, but it is more inclusive than some other communities around, women are all around the MonoGame community (i would know, being a woman and all), if you want to be a negative nancy, keep it out of the space where people want to get work done.

**Comment by damian-666** [comment]
its not my point to insult anyone its to try and improve the culture in a way to protect and prevent the insulting of other people coming in here tryying to help, learn, or to prototype , so that they can MVP or to   convince VC's to back their projects that rely heavily on this thing.  those VCs might contribute large sums   .   but if this toxicity goes on i can't justify continued advocation for  my partners investments. 

 .. i just looked at how many of my comments were collapsed and reread them.  ive see how vector2<int>   PR was treated..this sdeserves some respect  ( hes from a Quantum state collapse vault ) but  he  was dissed and sent off.    Its true hsi PR should not be taken  at present but he deserves a modicum of respect.  Someone put a lauging face when he was hurt , closed PR and left.  laughing at someone in pain and frustration?   that is pure TOXIC.    it hurt me to see that. people have feelings and theres so much  hurt you can take before you lash out.
 
  Some were sooo positive i was just shocked..  some were long and offtopic but if peope  lack the imagination to cant connec the dots, its not my fault.  they can skip it.   i dont mind being insulted but as Voltaire said no one ahs the right to censor.  I'd rather be insulted than deleted after i almost DIED doinga web based game deployment and shut down by chrome... They just are too vulnerable because people are vulnerable and the web is too easy to access. So they will get blocked.  There are no stars, no common  rating system, no way to protect the users, and hard to make an revenue model.  performance = security risk.  if you dont have accountability its going very temporary.    you dont see nkast lately around for a reason ,hes been disrespected in here, and deserves better.  Anwyas monogame saved my life, gave me a second chance.. but the culture could be greatly improved.  
  
    merge  some more PRs , let people speak, and close issues you'll get your patronage.   this not a tight ship.    Negavity is not a bad thing, if you embrace every tech  you sink with all those ships.        but respect is paramount.    anyways i have what i need,  I won't raise any issues and I think the project is going well despite being nearly impossible to maintain ,  its a miracle.   I have taken the pain, i can and will take more, but I just had to say something about the inclusiveness and the toxicity towards others ive seen,  i certainly don't mean to defend mysefl or  make it more toxic or be a Negative Creep all the time.  thx for reading and blessings to you.

**Comment by Beyley** [comment]
You say your point is not to insult, yet you say things like "bunch of grumpy old dudes" and "the best most of you are gonig to get from a real female... LOL."
You aren't going to win anyone's heart by insulting them, then claiming its to 'protect people'(?)
If you want to foster a positive community, be positive, being negative helps no one, especially in a professional setting like GitHub (also can you point to a place in here where people have disrespected nkast? ive only seen neutral talk/praise for their work in this thread)

**Comment by SimonDarksideJ** [comment]
Closing in favour of https://github.com/MonoGame/MonoGame/pull/9137, which is closer to a working solution.  Please reopen the discussion if needed or continue there.
